File: src/journal/sim/vwap.py
"""Session-anchored VWAP and deviation bands, computed tick by tick.

This is *the* VWAP definition in this codebase, and the reason it is worth saying
so: accumulating trade prices and accumulating 1-minute bar typical prices are
two different statistics, not two spellings of one. Their mids differ by a tick
or three (hlc3 is only a proxy for a bar's true VWAP) and their sigmas differ by
far more near an anchor — by the law of total variance the tick sigma carries the
within-bar spread as well as the between-bar one, so it is ~40x wider one bar in,
and converges to the bar figure only some tens of bars later. Drawing both under
one name is what made a hand-placed anchor read differently on two charts.

So the bands the engine trades against, the session lines the charts draw, and
every cumulative anchored VWAP on the frontend all come from here or from
``bar_moments`` below — a strategy tested on one sigma and shown on another is
untestable, and a tool that means two things is worse than either.

Volume-weighted, which is the standard (and what ATAS draws):

    vwap = sum(p*v) / sum(v)
    var  = sum(p^2*v) / sum(v) - vwap^2
    dev1 = vwap +/- sqrt(var),  dev2 = vwap +/- 2*sqrt(var)
"""

from __future__ import annotations

import numpy as np
import pandas as pd

BAND_COLS = ["mid", "std", "upper1", "upper2", "lower1", "lower2"]


def vwap_bands(ticks: pd.DataFrame,
               seed: tuple[float, float, float] | None = None) -> pd.DataFrame:
    """Running VWAP + 1σ/2σ bands, one row per tick, index-aligned to *ticks*.

    Accumulation starts at the first row, so the caller anchors the session by
    slicing the tick frame (e.g. from the 09:30 ET open) before calling. An
    anchor that predates the frame — the weekly VWAP, anchored at the week's
    Globex open but computed over one session's ticks — is expressed as *seed*:
    the (Σv, Σpv, Σp²v) already accumulated between the anchor and the frame's
    first tick (see ``weekly.weekly_seed``).

    The first few ticks have a near-zero sigma — the bands are degenerate until
    some volume has traded. That is honest rather than a bug: it is exactly what
    a live VWAP looks like seconds after the open. Rules must not lean on them.
    (A seeded call starts with the anchor's volume behind it, so its bands are
    real from the first tick.)
    """
    if ticks.empty:
        return pd.DataFrame(columns=BAND_COLS)

    p = ticks["price"].to_numpy(dtype="float64")
    v = ticks["size"].to_numpy(dtype="float64")

    sv, spv, sp2v = seed if seed is not None else (0.0, 0.0, 0.0)
    cum_v = sv + np.cumsum(v)
    cum_pv = spv + np.cumsum(p * v)
    cum_p2v = sp2v + np.cumsum(p * p * v)

    with np.errstate(divide="ignore", invalid="ignore"):
        mid = cum_pv / cum_v
        var = cum_p2v / cum_v - mid * mid
    # Catastrophic cancellation can push a true-zero variance a hair negative.
    std = np.sqrt(np.clip(var, 0.0, None))

    return pd.DataFrame({
        "mid": mid,
        "std": std,
        "upper1": mid + std,
        "upper2": mid + 2 * std,
        "lower1": mid - std,
        "lower2": mid - 2 * std,
    }, index=ticks.index)


def bar_moments(ticks: pd.DataFrame,
                start_idx: np.ndarray,
                end_idx: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Each bar's own tick VWAP and tick variance, one pair per bar.

    This is what lets a chart draw a *tick-accurate* VWAP anchored anywhere
    without being shipped the tape. Sums are associative, so accumulating these
    forward from any bar reproduces ``vwap_bands`` over the same tick range
    exactly — the frontend rebuilds (Σv, Σpv, Σp²v) from ``(volume, vwap, var)``
    as ``pv = vwap·v`` and ``p²v = (var + vwap²)·v``. Two floats per bar buys
    every anchor the user can place, at bar cost rather than tick cost. Same
    trick ``frame_sums`` plays for the weekly seed, one bar wide.

    ``start_idx``/``end_idx`` are positional into *ticks*, inclusive at both
    ends, exactly as ``bars.BAR_COLS`` carries them.

    The variance is accumulated *about each bar's own mean* rather than as
    E[x²]−E[x]², because the difference-of-cumulative-sums form loses most of its
    significant digits here: p² is ~6e8 on NQ and a bar's true variance is single
    digits, so the naive route subtracts two nearly equal large numbers to get a
    small one. Centring costs one extra pass and is exact.

    NaN for a bar that caught no volume — a time bar over a hole. Callers drop
    those rather than drawing a made-up price; a zeroed moment would silently
    drag an anchored VWAP toward zero.
    """
    n = len(start_idx)
    if ticks.empty or n == 0:
        return np.full(n, np.nan), np.full(n, np.nan)

    p = ticks["price"].to_numpy(dtype="float64")
    v = ticks["size"].to_numpy(dtype="float64")
    s = np.asarray(start_idx, dtype="int64")
    e = np.asarray(end_idx, dtype="int64")

    # Which bar each tick belongs to. `end_idx` is inclusive and ascending, so
    # side="left" lands a tick on the first bar whose end it does not pass —
    # the same mapping session_chart builds for its footprint and CVD lookups.
    bar_of = np.searchsorted(e, np.arange(len(p)), side="left")
    # Ticks before the first bar opens, or past the last one's close, belong to
    # no drawn bar. Both happen: an NY-anchored slice starts mid-frame, and a
    # session can carry recovered ticks past the final bar.
    live = (np.arange(len(p)) >= s[0]) & (bar_of < n)
    bo, pv_p, pv_v = bar_of[live], p[live], v[live]

    vol = np.bincount(bo, weights=pv_v, minlength=n)
    pv = np.bincount(bo, weights=pv_p * pv_v, minlength=n)
    with np.errstate(divide="ignore", invalid="ignore"):
        mid = pv / vol
    d = pv_p - mid[bo]
    var = np.bincount(bo, weights=d * d * pv_v, minlength=n)
    with np.errstate(divide="ignore", invalid="ignore"):
        var = var / vol

    empty = vol <= 0
    mid = np.where(empty, np.nan, mid)
    var = np.where(empty, np.nan, np.clip(var, 0.0, None))
    return mid, var


def frame_sums(ticks: pd.DataFrame) -> tuple[float, float, float]:
    """(Σv, Σpv, Σp²v) over a whole tick frame — one session's contribution to a
    multi-session seed. Summing these across days and passing the total as
    ``vwap_bands(seed=...)`` is exactly equivalent to accumulating over the
    concatenated tick frames."""
    if ticks.empty:
        return (0.0, 0.0, 0.0)
    p = ticks["price"].to_numpy(dtype="float64")
    v = ticks["size"].to_numpy(dtype="float64")
    return (float(v.sum()), float((p * v).sum()), float((p * p * v).sum()))

File: frontend/src/lib/vwap.ts
// The volume-weighted accumulator, and what a bar contributes to one.
//
// The frontend half of `src/journal/sim/vwap.py`, and the same rule applies on
// this side of the wire: **there is one VWAP in this app and it accumulates
// trade prices.** Accumulating bar typical prices instead is a different
// statistic, not a cheaper spelling of the same one — the mids differ by a tick
// or three, and near an anchor the sigmas differ by far more, because a tick
// sigma carries the spread *inside* each bar as well as the spread between them.
// Drawing both under one name is what let a hand-placed ⚓ read differently on
// two charts, which is the bug this module exists to make unrepeatable.
//
// Two ways in, one arithmetic:
//
//   * **Tick by tick** — `add(price, size)`, which is what the replay engine
//     does as the tape plays.
//   * **Bar by bar** — `addBar(bar)`, for a chart that was never shipped the
//     tape. Each bar carries its own tick VWAP and tick variance (`tvwap`,
//     `tvar`, from `journal.sim.vwap.bar_moments`), and because sums are
//     associative, folding those forward from any anchor reproduces the tick
//     accumulation over the same range *exactly* — verified bit-exact on the
//     mid and to 8e-8 on sigma against a real 47k-tick session. Two floats per
//     bar buys every anchor the user can place, at bar cost rather than tick
//     cost.
//
// A bar with no moments falls back to `hlc3 × volume`, which is the old
// bar-domain approximation and is *not* interchangeable with the rest. Only one
// payload still lacks them (`api/charts_data.bars_window`, the timeframe-radio
// refetch, which is resampled from the bar store and has no tape behind it), and
// callers that can tell the user which arithmetic they got should.

/** The shape `addBar` reads. Structural, so both charts' `Bar` types satisfy it
 *  without either importing the other — same reason lib/modernVwap declares its
 *  own input shape. */
export interface VwapBar {
  high: number;
  low: number;
  close: number;
  volume: number;
  /** This bar's own tick VWAP, `Σ(p·v)/Σv` over its ticks. Absent where no tape
   *  stood behind the bar, or where the bar caught no volume at all. */
  tvwap?: number;
  /** This bar's own tick variance about `tvwap`. Absent alongside it. */
  tvar?: number;
}

/** Whether a bar can contribute tick-accurate moments, or only the hlc3
 *  approximation. Charts quote this so the two are never silently mixed. */
export function hasTickMoments(b: VwapBar): boolean {
  return Number.isFinite(b.tvwap) && Number.isFinite(b.tvar);
}

/** One bar's `(v, pv, p²v)`, tick-exact where the moments are there and hlc3 as
 *  the honest fallback where they are not. */
export function barMoments(b: VwapBar): { v: number; pv: number; p2v: number } {
  const v = b.volume > 0 ? b.volume : 0;
  if (!v) return { v: 0, pv: 0, p2v: 0 };
  if (hasTickMoments(b)) {
    const m = b.tvwap as number;
    // `p²v` back out of the variance: E[x²] = Var + E[x]². Shipping the variance
    // rather than E[x²] is what keeps the wire numbers small and the precision
    // good — see bar_moments' note on centring.
    return { v, pv: m * v, p2v: ((b.tvar as number) + m * m) * v };
  }
  const tp = (b.high + b.low + b.close) / 3;
  return { v, pv: tp * v, p2v: tp * tp * v };
}

/** A bar being built tick by tick, for `openBar`/`foldTick` below. */
interface FormingBar {
  volume: number;
  tvwap: number;
  tvar: number;
}

/** The moment fields a bar opens with: one observation, so the bar's VWAP is
 *  that print and its spread is nothing. */
export function openBar(price: number): { tvwap: number; tvar: number } {
  return { tvwap: price, tvar: 0 };
}

/**
 * Fold one print into a forming bar — volume and both moments together.
 *
 * One function rather than a `volume +=` beside two more lines at each of the
 * three places a bar grows: a bar whose volume and moments disagree would draw a
 * VWAP that is wrong by an amount nothing else on the chart could explain.
 *
 * `E[x²] = Var + E[x]²` on the way in and back out again. Within a single bar
 * the round trip is harmless — the sums are one bar's worth, so the subtraction
 * keeps ~8 significant digits on a variance of single digits. It is only across
 * a whole session that this form collapses, which is exactly why the *shipped*
 * per-bar variance is centred instead (see `journal.sim.vwap.bar_moments`).
 */
export function foldTick(bar: FormingBar, price: number, size: number): void {
  const v0 = bar.volume;
  const v1 = v0 + size;
  if (!(v1 > 0)) {
    bar.volume = v1;
    return;
  }
  const pv = bar.tvwap * v0 + price * size;
  const p2v = (bar.tvar + bar.tvwap * bar.tvwap) * v0 + price * price * size;
  const m = pv / v1;
  bar.volume = v1;
  bar.tvwap = m;
  bar.tvar = Math.max(p2v / v1 - m * m, 0);
}

/** A point on an anchored VWAP: the mid and both deviation bands. */
export interface VwapBandPoint {
  mid: number;
  sd: number;
  u1: number;
  l1: number;
  u2: number;
  l2: number;
}

/**
 * Volume-weighted accumulator for one anchored VWAP.
 *
 * `seed` is what was already behind the anchor before the first observation this
 * accumulator will ever see — the weekly anchor, which starts days before the
 * tape (see `journal.sim.vwap`'s `seed`). Zero for an anchor that starts inside
 * the range, which is every other one.
 */
export class Vwap {
  v = 0;
  pv = 0;
  p2v = 0;

  constructor(seed?: readonly number[] | null) {
    if (seed && seed.length === 3) {
      this.v = seed[0];
      this.pv = seed[1];
      this.p2v = seed[2];
    }
  }

  add(price: number, size: number) {
    this.v += size;
    this.pv += price * size;
    this.p2v += price * price * size;
  }

  /** Fold in a whole bar. A bar with no volume moves nothing — it must not be
   *  allowed to drag the average, which is what a zeroed moment would do. */
  addBar(b: VwapBar) {
    const m = barMoments(b);
    if (!m.v) return;
    this.v += m.v;
    this.pv += m.pv;
    this.p2v += m.p2v;
  }

  /** Mid and sigma as they stand. NaN before anything has traded under the
   *  anchor — a caller draws a gap there rather than inventing a price. */
  read(): { mid: number; sd: number } {
    if (!(this.v > 0)) return { mid: NaN, sd: NaN };
    const mid = this.pv / this.v;
    // Clamped: the two moments accumulate independently, so rounding can put
    // E[x²] a hair under E[x]² on a range that has barely moved, and a negative
    // variance would come back NaN from the sqrt. Same guard as vwap.py's.
    return { mid, sd: Math.sqrt(Math.max(this.p2v / this.v - mid * mid, 0)) };
  }

  /** The full band, for the callers that draw all of it. */
  band(): VwapBandPoint {
    const { mid, sd } = this.read();
    return { mid, sd, u1: mid + sd, l1: mid - sd, u2: mid + 2 * sd, l2: mid - 2 * sd };
  }

  get active() {
    return this.v > 0;
  }
}
